lumicks/pylake/population/dwelltime.py

```diff
@@ -761,6 +768,60 @@ def label_maker(a, t, n):
         plt.tight_layout()
 
 
+def _exponential_mixture_log_likelihood_gradient(
+    params, t, min_observation_time, max_observation_time
+):
+    """Calculate the gradient of the exponential mixture model.
+
+    Parameters
+    ----------
+    params : array_like
+        array of model parameters (amplitude and lifetime per component)
+    t : array_like
+        dwelltime observations in seconds
+    min_observation_time : float
+        minimum observation time in seconds
+    max_observation_time : float
+        maximum observation time in seconds
+    """
+    amplitudes, lifetimes = np.reshape(params, (2, -1))
+    # We clip the amplitude lower bound. An amplitude of 1e-14 is negligible.
+    amplitudes = np.clip(amplitudes[:, np.newaxis], 1e-14, np.inf)
+    lifetimes = lifetimes[:, np.newaxis]
+    t = t[np.newaxis, :]
+
+    t_min, t_max = min_observation_time, max_observation_time
+
+    exp_bound_difference = np.exp(-t_min / lifetimes) - np.exp(-t_max / lifetimes)
+    norm_factor = 1.0 / np.sum(amplitudes * exp_bound_difference)
+
+    # Calculate derivatives of the normalization constant
+    dnorm_damp = -(norm_factor**2) * exp_bound_difference
+    dlognorm_damp = (1.0 / norm_factor) * dnorm_damp
+
+    max_bound = t_max * np.exp(-t_max / lifetimes) if np.all(t_max / lifetimes < 1e10) else 0
+    exp_bound_difference2 = max_bound - t_min * np.exp(-t_min / lifetimes)
+    dnorm_dtau = norm_factor**2 * (amplitudes / (lifetimes**2)) * exp_bound_difference2
+    dlognorm_dtau = (1.0 / norm_factor) * dnorm_dtau
+
+    # Calculate derivatives of the remainder of the log-likelihood
+    dlogamp_damp = 1.0 / amplitudes
+    dlogtauterm_dtau = -1.0 / lifetimes + t / (lifetimes**2)
+
+    # The derivative of logsumexp is given by: sum(exp(fi(x)) dfi(x)/dx) / sum(exp(fi(x)))
+    components = np.log(norm_factor) + np.log(amplitudes) - np.log(lifetimes) - t / lifetimes
+    total_denom = np.exp(logsumexp(components, axis=0))
+
+    sum_components = np.sum(np.exp(components), axis=0)
+    dtotal_damp = (sum_components * dlognorm_damp + np.exp(components) * dlogamp_damp) / total_denom
+    dtotal_dtau = (
+        sum_components * dlognorm_dtau + np.exp(components) * dlogtauterm_dtau
+    ) / total_denom
+    unsummed_gradient = np.vstack((dtotal_damp, dtotal_dtau))
+
+    return -np.sum(unsummed_gradient, axis=1)
+
+
 def exponential_mixture_log_likelihood_components(
```

Comment:
I think for "public" functions like these it might be a good time to rename them with the leading `_` and denote them as internal. I don't think this is a breaking change since we don't expose these through the package `__init__.py` and the module is specifically marked alpha functionality; more of a safeguard for our future selves